Add exercise for a Basic OCI setup that will be used for the OCI Guide (#21)
helidon-io/helidon-labs: hols/oci-basic-setup/README.md

> # Deploying Helidon OCI MP Application on a Basic OCI Infrastructure Setup

Author: Updated

> This example will demonstrate how to create a basic OCI infrastructure setup that can be used to deploy a Helidon MP OCI application.

Reviewer: ```suggestion
This example demonstrates how to create a basic OCI setup that can be used to deploy a Helidon MP OCI application.
```

Reviewer: We try to use present tense over future whenever possible.

## Objective

> 1. Use Terraform to create automation of OCI resources provisioning that can build a basic OCI infrastructure setup composed of the following:

Author: Updated

>    1. OCI Compute Instance with firewall opened at port 8080.

Author: updated

>    2. OCI Virtual Cloud Network with Security List containing an Ingress at port 8080.

Author: updated

   3. Policies to allow OCI Logging and Monitoring services to be accessed from the provisioned OCI Compute Instance.
2. Generate a Helidon MP OCI project using Helidon cli. 

> 3. Use OCI cloud-shell to run this example through. 

Author: Updated

## Prerequisite

> - An OCI tenancy that has enough capacity to provision an OCI Compute Instance and a Virtual Cloud Network (VCN). This will also work on the free trial of the [OCI Free Tier](https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier.htm).

Author: Is this ok:

An OCI tenancy that has enough capacity to provision an OCI Compute Instance,  a Virtual Cloud Network (VCN) and a Custom Log. At a minimum, this will also work on an [Always Free Resources](https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier_topic-Always_Free_Resources.htm) type of Tenancy.

> 

Reviewer 2: Consider adding some text that reminds the reader that this is not a tutorial on OCI, OCI GUI, or OCI configuration and services. We want them to learn that from OCI provided resources.
Perhaps describe the basic actions they must have completed prior to performing any of the tasks in this example.
If the reader has never used OCI, they could read [Signing In for the First Time](https://docs.public.oneportal.content.oci.oraclecloud.com/en-us/iaas/Content/GSG/Tasks/signingin_topic-Signing_In_for_the_First_Time.htm)
If there is a basic tutorial or HOL from OCI on creating a first OCI account, maybe provide a reference. This tutorial uses Resource Manager may be worth mentioning (I don't know if this is the best one, or not) [Quickly Launch Your Always Free Resources Using Resource Manager](https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier_topic-Always_Free_Resources_Launching.htm)
I think much of what is performed in this example is described in the OCI Tutorial [Launching your first Linux instance](https://docs.oracle.com/en-us/iaas/Content/Compute/tutorials/first-linux-instance/overview.htm) -- this could be useful for users that want to learn more about OCI
I think you already have a reference to the OCI page for [Always Free Resources](https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier_topic-Always_Free_Resources.htm)

Author: Updated

> - An OCI tenancy that has enough capacity to provision an OCI Compute Instance and a Virtual Cloud Network. This will also work on the free trial of the [OCI Free Tier](https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier.htm).

Author: Good suggestion, updated

## Tasks
### Set up Cloud Shell access.

> 1. If you are already a super-user with administrator rights that has access to all resources in the tenancy, skip to the next section. Otherwise, continue to the next step.

Author: Both should be the same from policy perspective.

> 1. If you are already a super-user with administrator rights that has access to all resources in the tenancy, skip to the next section. Otherwise, continue to the next step.

Reviewer 3: If they are not a super/admin user than they wouldn't have rights, to setup below policies, correct? May be, we need to say, make sure your admin has given access to you or ask them to setup following for your user.

2. Create a new group and add your user as a member of that group.
3. Create a new policy and provide that group a cloud-shell access:
   ```
   Allow group '<my_cloud_shell_access_group>' to use cloud-shell in tenancy
   ```
4. Verify that it works by opening Cloud Shell from the OCI Console.

### Retrieve the oci-basic-setup subdirectory from the helidon-labs repository

> The materials for this exercise will be located in the [oci-basic-setup](https://github.com/helidon-io/helidon-labs/tree/main/hols/oci-basic-setup) subdirectory of the  [helidon-labs](https://github.com/helidon-io/helidon-labs) repository, hence only that specific subdirectory will be cloned.

Author: Updated

1. Open a Cloud Shell terminal and make sure you are in the home directory
   ```shell
   cd ~
   ```
2. Use `git sparsecheckout` to check out only the `oci-basic-setup` directory from the helidon-labs repository.
   ```shell
   git init helidon-labs
   cd helidon-labs 
   git remote add origin https://github.com/helidon-io/helidon-labs.git
   git config core.sparsecheckout true
   echo "hols/oci-basic-setup/*" >> .git/info/sparse-checkout
   echo ".gitignore" >> .git/info/sparse-checkout
   git pull --depth=1 origin main
   ```

> 3. The previous step will pull all the required Terraform and Bash script files into the  `~/helidon-labs/hols/oci-basic-setup` directory that will be needed to perform various operations to complete this exercise.

Author: updated

### Prepare the OCI infrastructure environment
The goal of this task is to prepare a basic infrastructure environment comprised of a Compartment, Dynamic Groups, Policies, Compute and Virtual Cloud Network. This section requires a user with administrator privilege.
1. Open the newly cloned repository directory `oci-basic-setup` from the a Cloud Shell terminal.
   ```shell
   cd ~/helidon-labs/hols/oci-basic-setup
   ```

> 2. From the root directory of the repository, open `terraform.tfvars` and set the values of the following variables: 

Author: updated

>    * `tenancy_ocid` - This can be retrieved by opening the `Profile` menu (click the User icon on the top most right corner of the console window) and click `Tenancy: <your_tenancy_name>`. The tenancy OCID is shown under `Tenancy Information`. Click `Show` to display the entire ID or click `Copy` to copy it to your clipboard.

Author: Updated

   * `region` - From OCI Console's home screen, open the `Region` menu, and then click `Manage Regions`. Locate the Home Region and copy the Region identifier.

>    * `instance_*` - Use these parameters to specify the size and shape of the OCI Compute instance that you would like to be provisioned. Currently defaults to an available shape from an [Always Free Resources of an OCI Free Tier Tenancy](https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier_topic-Always_Free_Resources.htm).

Author: Updated

>    * `instance_*` - Use these parameters to specify the size and shape of the OCI compute instance that you would like to be provisioned. Currently defaults to an available shape from an [Always Free Resources of an OCI Free Tier Tenancy](https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier_topic-Always_Free_Resources.htm).

Author: updated

     ```terraform
     instance_shape                      = "VM.Standard.E2.1.Micro"
     instance_ocpus                      = 1
     instance_shape_config_memory_in_gbs = 1
     instance_os                         = "Oracle Linux"
     instance_os_version                 = "8"
     ```
4. Use Terraform to execute the scripts to provision the OCI resources.
   ```bash
   terraform init
   terraform plan
   terraform apply -auto-approve
   ```
    
### Prepare the Helidon OCI Application:
1. Go to the home directory from a Cloud Shell console:
   ```shell
   cd ~
   ```  

> 2. Download and unzip the helidon cli generic distribution

Author: updated

   ```shell
   curl -L -O https://github.com/helidon-io/helidon-build-tools/releases/download/3.0.6/helidon-cli.zip

>    unzip helidon-cli.zip

Reviewer 2: Consider adding ~/helidon-3.0.6/bin to PATH (and maybe a suggestion to add this to ~/.bashrc).
Note, this may be a maintenance issue as it will change whenever a new version of the cli is created.

Author: Updated

   ```

> 3. Make sure that JDK 21 exists in the path.

Reviewer 2: (run java -version, or mvn -v -- Provide a reference to resolve this?

Author: Updated

> 3. Make sure that JDK 21 exist in the path.

Author: updated

   ```shell
   Helidon versions
   (1) 4.2.1
   (2) 3.2.12
   (3) 2.6.11
   (4) Show all versions
   Enter selection (default: 1):
   ```
6. When prompted to `Select a Flavor`, choose `Helidon MP`
   ```shell
   | Helidon Flavor
   
   Select a Flavor
     (1) se | Helidon SE
     (2) mp | Helidon MP
   Enter selection (default: 1): 2
   ```
7. When prompted to `Select an Application Type`, choose `oci`  
   ```shell
   Select an Application Type
   (1) quickstart | Quickstart
   (2) database   | Database
   (3) custom     | Custom
   (4) oci        | OCI
   Enter selection (default: 1): 4
   ```
8. When prompted for `Project groupId`, `Project artifactId`, `Project version` and `Java package name`, just accept the default values. Once completed, this will generate an `oci-mp` project.
10. Go to the `oci-mp` directory.
    ```shell
    cd ~/oci-mp
    ```
11. Run the `update_config_values.sh` utility script from the `oci-basic-setup` local repository to update the config parameters. Specify the directory location of the Helidon application as an argument to this script.
    ```shell
    ~/helidon-labs/hols/oci-basic-setup/update_config_values.sh ~/oci-mp
    ```

>     Invoking this script will perform the following:

Author: Updated

>     1. Updates in `~/oci-mp/server/src/main/resources/application.yaml` config file to set up a Helidon feature that sends Helidon generated metrics to the OCI monitoring service.

Reviewer: ```suggestion
    1. Updates the `~/oci-mp/server/src/main/resources/application.yaml` config file to set up a Helidon feature that sends Helidon generated metrics to the OCI Monitoring service.
```

Author: Updated

>        1. compartmentId - Compartment ocid that is used for this demo

Reviewer: ```suggestion
       1. compartmentId - The Compartment OCID used for this demo
```

Author: Updated

       2. namespace - This can be any string but for this demo, will be set to `helidon_metrics`.

>     2. Updates in `~/oci-mp/server/src/main/resources/META-INF/microprofile-config.properties` config file to set up configuration parameters used by the Helidon app code to perform integration with OCI Logging and Metrics service.

Reviewer: ```suggestion
    2. Updates the `~/oci-mp/server/src/main/resources/META-INF/microprofile-config.properties` config file to set up configuration parameters used by the Helidon application code to perform integration with OCI Logging and Metrics service.
```

Author: Updated

       1. oci.monitoring.compartmentId - Compartment ocid that is used for this demo
       2. oci.monitoring.namespace - This can be any string but for this demo, this will be set to `helidon_application`.
       3. oci.logging.id - Application log id that was provisioned by the terraform scripts.

> 

Reviewer 2: Add comment that the user needs to run update_config_values.sh whenever they recreate their compute instance. (ref. the demo you and I did earlier today)

Author: Updated troubleshooting tips for this.

>     **Note:** Make sure to validate that `application.yaml` and `microprofile-config.properties` have been updated by checking that the mentioned config parameters were properly populated.

Author: updated

> 12. To prepare for the build, ensure that JDK 21 and Maven 3.8+ exist and are set in the PATH environment variable of the Cloud Shell terminal.

Reviewer 2: Run mvn -v and confirm the java version as well as the maven version

Author: Updated by removing this step, as prior steps already describes how check/download/install java and mvn

13. Build the application.  
    ```shell
    mvn clean package
    ```

> ### Deploy the Helidon OCI application:

Reviewer 2: Deploy the Helidon OCI Application to your compute instance

Author: Updated

1. Go back to `oci-basic-setup` local repository directory.
   ```shell
   cd ~/helidon-labs/hols/oci-basic-setup
   ```

> 2. Open up `deploy.sh` script and use the JDK_TAR_GZ_INSTALLER variable to specify the download path of jdk 21 based on the OS flavor currently installed on Compute instance. The current value used is the Linux x64 version of jdk 21. 

Author: updated

3. Run `deploy.sh` utility script and provide the directory of the OCI MP application as the argument. This script will zip up the server binaries, upload it to the Compute instance, install JDK 21, create a systemctl service for the application to ensure that the application is persistent across reboots, and start the service.
   ```shell
   ./deploy.sh ~/oci-mp
   ```

### Exercise the deployed Helidon application:
1. Access the application by using curl to perform GET & PUT HTTP requests.
    1. Set the endpoint using the instance's public ip which can be retrieved using `get.sh` utility script from the `oci-basic-setup` local repository directory.
       ```shell
       export ENDPOINT_IP=$(~/helidon-labs/hols/oci-basic-setup/get.sh public_ip)
       echo "Instance public ip is $ENDPOINT_IP"
       ```

>     2. Test the Hello world request.

Reviewer 2: after this line maybe include: "These are some commands you can invoke on the Helidon application that you just deployed:"

Author: I think the step descriptuion indicating `Access the application by using curl to perform GET & PUT HTTP requests.` should be enough to convey the same message.

>     2. Test Hello world request.

Author: updated

       ```shell
       curl http://$ENDPOINT_IP:8080/greet
       ```
       results to:
       ```shell
       {"message":"Hello World!","date":[2025,4,29]}
       ```
    3. Test Hello to a name, i.e. to `Joe`.
       ```shell
       curl http://$ENDPOINT_IP:8080/greet/Joe
       ```

>        results to:

Reviewer: Global change

   2. On `Query 1`, choose the `Compartment` value with the format of `devops-compartment-helidon-demo-<4 char random value>`.
   3. Select `helidon_metrics` under `Metric namespace`.
   4. Select `requests.count_counter` under `Metric Name`.
   5. Above the empty graph, you can choose values for `Start time/End Time` or choose the time duration under `Quick Selects`.
   6. Click `Update Chart` at the bottom of `Query 1` to show all the metric data for the `request count`. You can also toggle to enable `Show Data Table` on the right upper portion of the graph to show the list of data for the chosen metric.
   7. You can also explore other Metrics by going back to step 4 and choosing a new value. 
3. Validate OCI Logging SDK integration that was added in the Helidon application. This will push log messages to the OCI Logging Service:
   1. From the OCI Console, go to `Observability & Management` -> `Logs (under Logging)`.
   2. Change `Compartment` with a value that has the format of `devops-compartment-helidon-demo-<4 char random value>`.
   3. Under `Filters`, change `Log Group` to `app-log-group-helidon-demo`.
   4. Choose and click on `app-log-helidon-demo` from the Logs table.

>    5. Choose `Filter by time` value within the scope of your last request. For example, you can choose `Today` to see all request that was made today.

Author: updated

>    6. The `Explore Log` display should output some graphs of the logging activity and below it will also show a list of the logs that has been captured.

Author: updated

### Cleanup
When the environment is no longer needed, all the OCI resources can be cleaned up by following these steps:
1. Go back to oci-basic-setup local repository directory.
   ```shell
   cd ~/helidon-labs/hols/oci-basic-setup
   ```  
2. Execute the `destroy` option from terraform.
   ```shell
   terraform destroy -auto-approve

>    ```   

Reviewer 2: Consider adding a troubleshooting section as we discussed

Author: Updated

# Deploying a Helidon OCI MP Application on a Basic OCI Setup

> This example demonstrates how to create a basic OCI infrastructure setup that can be used to deploy a Helidon MP OCI application.

Author: I termed it `infrastructure` because it is a collection of OCI resources and Compure is just a part of it

## Objective
1. Use Terraform to automate the provisioning of OCI resources that can build a basic OCI setup composed of the following:
   1. OCI Compute instance with firewall opened at port 8080.
   2. OCI Virtual Cloud Network with a Security List containing an Ingress at port 8080.
   3. Policies to allow OCI Logging and Monitoring services to be accessed from the provisioned OCI Compute instance.

> 2. Generate a Helidon MP OCI project using Helidon cli. 

Author: Updated

   ```
4. Verify that it works by opening Cloud Shell from the OCI Console.

### Retrieve the oci-basic-setup subdirectory from the helidon-labs repository
The materials for this exercise are located in the [oci-basic-setup](https://github.com/helidon-io/helidon-labs/tree/main/hols/oci-basic-setup) subdirectory of the  [helidon-labs](https://github.com/helidon-io/helidon-labs) repository, so only that specific subdirectory needs to be cloned.
1. Open a Cloud Shell terminal and make sure you are in the home directory
   ```shell
   cd ~
   ```
2. Use `git sparsecheckout` to check out only the `oci-basic-setup` directory from the `helidon-labs` repository.
   ```shell
   git init helidon-labs
   cd helidon-labs 
   git remote add origin https://github.com/helidon-io/helidon-labs.git
   git config core.sparsecheckout true
   echo "hols/oci-basic-setup/*" >> .git/info/sparse-checkout
   echo ".gitignore" >> .git/info/sparse-checkout
   git pull --depth=1 origin main
   ```
3. The previous step pulls all the required Terraform and Bash script files into the  `~/helidon-labs/hols/oci-basic-setup` directory that is required to perform the various operations in this exercise.

> ### Prepare the OCI infrastructure environment

Author: I termed it `infrastructure` because it is a collection of OCI resources and Compure is just a part of it

      ```
      **Note:** If you want the above environment variable settings to persist between Cloud Shell sessions, add the definitions to your `~/.bashrc` file.
3. Make sure that Maven 3.8 or higher exists.
    1. Run the following command to validate:
       ```shell
       mvn --version
       ```
    2. If it doesn't exist, download and install the binary archive of your choice from the [Apache Maven Download page](https://maven.apache.org/download.cgi). Below is an example of downloading and extracting a  Maven 3.9.9 binary tar.gz archive.
       ```shell
       curl -O https://dlcdn.apache.org/maven/maven-3/3.9.9/binaries/apache-maven-3.9.9-bin.tar.gz
       tar -xvzf apache-maven-3.9.9-bin.tar.gz
       ```
    3. Add the extracted mvn bin subdirectory into the PATH environment variable
       ```shell
       export PATH=~/apache-maven-3.9.9/bin/:$PATH
       ```
       **Note:** If you want the above environment variable setting to persist between Cloud Shell sessions, add the definition to your `~/.bashrc` file.
4. Install the Helidon CLI generic distribution.
    1. Download and extract the Helidon CLI generic distribution
       ```shell

>        curl -L -O https://github.com/helidon-io/helidon-build-tools/releases/download/3.0.6/helidon-cli.zip

Reviewer 3: The recommended approach to download CLI is `curl -L -O https://helidon.io/cli/latest/linux/helidon` based on https://helidon.io/docs/v4/about/cli

Author: Unfortunately, that will only work for Linux x64 platform. Cloud Shell no runs on Linux ARM by default and we don't have a distro for it yet, so the generic java distro is the only option for now.

   ```shell
   ~/helidon-labs/hols/oci-basic-setup/update_config_values.sh ~/oci-mp
   ```
   Invoking this script performs the following:
   1. Updates the `~/oci-mp/server/src/main/resources/application.yaml` config file to set up a Helidon feature that sends Helidon generated metrics to the OCI monitoring service.
      1. compartmentId - The Compartment OCID used for this demo.
      2. namespace - This can be any string but for this demo, will be set to `helidon_metrics`.
   2. Updates the `~/oci-mp/server/src/main/resources/META-INF/microprofile-config.properties` config file to set up configuration parameters used by the Helidon app code to perform integration with OCI Logging and Metrics service.
      1. oci.monitoring.compartmentId - Compartment ocid that is used for this demo
      2. oci.monitoring.namespace - This can be any string but for this demo, this will be set to `helidon_application`.
      3. oci.logging.id - Application log id that was provisioned by the terraform scripts.

>    **Note:** Make sure to validate that `application.yaml` and `microprofile-config.properties` were updated by checking that the mentioned config parameters were properly populated.

Reviewer 3: My 2cents: just keep properties file and not yaml.
